EventPhases/explainer.md

```diff
@@ -0,0 +1,501 @@
+
+### Non-goals
+
+ - Subtree-scale control for read and write phasing. It will be challenging enough for developers to consider a global regime; fine-grained configuration may harm adoption as much as could enhance it.
+ - New, fully asynchronous style readback APIs (see Considered Alternatives below).
+
+## Proposal
+
+We propose to introduce separate read and write phases when interacting with the DOM. When this mode is enabled, interleaving reads and writes will no longer incidentally trigger global style recalculation and layout, but will either result in "stale" reads or exceptions.
+
+Today's phasing can be described as:
+
+ 1. Browser processing (style, layout, etc.)
+ 1. JavaScript event dispatch
+    1. Task execution
+    1. Microtask queues drained
+ 1. GOTO: 1
+
+At any point in JavaScript processing, the DOM can be mutated and the results of those modifications on styling for any element can be synchronously requested, forcing _at least_ the `style` and `layout` phases of the frame production process to execute.
+
+This is, conceptually, a `read/write` phase, in which both operations are allowed in the same task or microtask.
+
+To be minimally invasive to this model, we propose additions to event handling that define phases on either side of `read/write`, such that either `read` or `write` callbacks execute before or after the existing event delivery tasks. We pair it with APIs for scheduling in future phases. 
+
+The new phasing, conceptually, is:
+
+ 1. Browser processing (style, layout, etc.)
+ 1. JavaScript event dispatch
+    1. Read phase
+       1. Task execution
+       1. Microtask queues drained
+    1. Read/Write
+       1. Task execution
+       1. Microtask queues drained
+    1. Write
+       1. Task execution
+       1. Microtask queues drained
+ 1. GOTO: 1
+
+In each turn, the `read` phase happens strictly before the `write` phase. We expect that style readback in the `write` phase will be allowed, but simply always return dirty information.
+
+For example:
+
+```js
+let foo = document.getElementById("foo");
+foo.addEventListener(
+  // Event type
+  "click", 
+  // Handler
+  (evt) => {
+    // Synchronous style readbacks are fine here
+    let w = foo.innerWidth;
+    let h = foo.innerHeight;
+    console.log(w, h);
+
+    // Writes are not allowed, and instead must 
+    // be scheduled. This could be written many 
+    // ways, but we imagine the Scheduler API 
+    // might be a good place for it:
+    scheduler.postTask(
+      () => {
+        let styles = foo.styleAttributeMap;
+        styles.set("width",  CSS.px(w + 10));
+        styles.set("height", CSS.px(h + 10));
+        // ...
+      },
+      { phase: "write" }
+    );
+  },
+  // Options
+  { phase: "read" }
+);
+
+// And using the new `when()` observable style:
```

Comment:
Personally - I'd lean towards simplifying everything by not including such a bleeding edge example with observables.